src/unix/kerberos_unix.cc

```diff
@@ -153,109 +143,99 @@ NAN_METHOD(InitializeClient) {
     }
 
     KerberosWorker::Run(callback, "kerberos:InitializeClient", [=](KerberosWorker::SetOnFinishedHandler onFinished) {
-        // TODO: Manage client_state as a proper C++ class with a destructor + through shared pointers
-        gss_client_state* client_state = gss_client_state_new();
+        auto client_state = std::make_shared<gss_client_state>();
         gss_result result = authenticate_gss_client_init(
-            service.c_str(), principal.c_str(), gss_flags, NULL, mech_oid, client_state);
-
-        if (result.code == AUTH_GSS_ERROR) {
-            authenticate_gss_client_clean(client_state);
-            free(client_state);
-        }
+            service.c_str(), principal.c_str(), gss_flags, NULL, mech_oid, client_state.get());
 
         return onFinished([=](KerberosWorker* worker) {
-            Nan::HandleScope scope;
+            Napi::Env env = worker->Env();
             if (result.code == AUTH_GSS_ERROR) {
-                v8::Local<v8::Value> argv[] = {Nan::Error(result.message.c_str()), Nan::Null()};
-                worker->Call(2, argv);
+                worker->Call(std::initializer_list<napi_value>
+                    { Error::New(env, result.message).Value(), env.Null() });
                 return;
             }
 
-            v8::Local<v8::Value> argv[] = {Nan::Null(), KerberosClient::NewInstance(client_state)};
-            worker->Call(2, argv);
+            worker->Call(std::initializer_list<napi_value>
+                { env.Null(), KerberosClient::NewInstance(env, client_state) });
```

Comment:
I wonder if it's a little unclear to future readers that `NewInstance` will take ownership of `client_state`? I think what happens here is that you invoke the copy ctor of the shared_ptr by passing `client_state` here, and then a move ctor inside `NewInstance` where you `std::move` the value. I think if we `std::move` _here_, then we better convey the semantics (don't try to use this anymore, we've moved it) and reduce a copy. What do you think?

Reply:
Sure – it doesn’t make much of a difference, practically speaking, so if it seems clearer then let’s do it :+1: Pushed a change for this.